lib/os/queued_operation.c

```diff
@@ -0,0 +1,444 @@
+	if (priority == QUEUED_OPERATION_PRIORITY_PREPEND) {
+		prio = INT8_MIN;
+	} else if (priority == QUEUED_OPERATION_PRIORITY_APPEND) {
+		prio = INT8_MAX;
+	} else if (prio != priority) {
+		return -EINVAL;
+	}
+
+	op->notify.flags = (op->notify.flags & ~mask)
+			   | (mask & (prio << QUEUED_OPERATION_PRIORITY_POS));
+
+	return 0;
+}
+
+static inline void finalize_and_notify(struct queued_operation_manager *mgr,
+				       struct queued_operation *op,
+				       int res)
+{
+	async_notify_generic_callback cb = async_notify_finalize(&op->notify, res);
+
+	if (cb != NULL) {
+		mgr->vtable->callback(mgr, op, cb);
+	}
+}
+
+/* React to the completion of an onoff transition, either from a
+ * manager or directly.
+ *
+ * @param mgr the operation manager
+ * @param from either ST_STARTING or ST_STOPPING depending on transition direction
+ * @param res the transition completion value, negative for error.
+ */
+static void settle_onoff(struct queued_operation_manager *mgr,
+			 enum state from,
+			 int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	__ASSERT_NO_MSG(mgr->state == from);
+
+	if (res >= 0) {
+		if (from == ST_STARTING) {
+			trivial_start_and_unlock(mgr, key);
+			return;
+		}
+
+		/* Came from STOPPING so set to OFF, but automatically
+		 * initiate a new start if operations have been added.
+		 */
+		mgr->state = ST_OFF;
+		if (!sys_slist_is_empty(&mgr->operations)) {
+			start_and_unlock(mgr, key);
+			return;
+		}
+
+		k_spin_unlock(&mgr->lock, key);
+		return;
+	}
+
+	/* On transition failure mark service failed.  All unstarted
+	 * operations are unlinked and completed as a service failure.
+	 */
+	sys_slist_t ops = mgr->operations;
+	sys_slist_init(&mgr->operations);
+	mgr->state = ST_ERROR;
+
+	k_spin_unlock(&mgr->lock, key);
+
+	struct queued_operation *op;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&ops, op, node) {
+		finalize_and_notify(mgr, op, -ENODEV);
+	}
+}
+
+static void start_callback(struct onoff_service *srv,
+			   struct onoff_client *cli,
+			   void *user_data,
+			   int res)
+{
+	settle_onoff(user_data, ST_STARTING, res);
+}
+
+static void stop_callback(struct onoff_service *srv,
+
+			  struct onoff_client *cli,
+			  void *user_data,
+			  int res)
+{
+	settle_onoff(user_data, ST_STOPPING, res);
+}
+
+static void stop_and_unlock(struct queued_operation_manager *mgr,
+			    k_spinlock_key_t key)
+{
+	__ASSERT_NO_MSG(mgr->state == ST_IDLE);
+
+	if (mgr->onoff == NULL) {
+		mgr->state = ST_OFF;
+		k_spin_unlock(&mgr->lock, key);
+		return;
+	}
+
+	mgr->state = ST_STOPPING;
+
+	struct onoff_service *onoff = mgr->onoff;
+
+	k_spin_unlock(&mgr->lock, key);
+
+	struct onoff_client *cli = &mgr->onoff_client;
+	int rc = 0;
+
+	onoff_client_init_callback(cli, stop_callback, mgr);
+	rc = onoff_release(onoff, cli);
+
+	if (rc < 0) {
+		settle_onoff(mgr, ST_STOPPING, rc);
+	}
+}
+
+static void select_next_and_unlock(struct queued_operation_manager *mgr,
+				   k_spinlock_key_t key)
+{
+	bool loop = false;
+
+	do {
```

Comment:
i still don't understand why you need a loop here. I only see a benefit that if process function is synchronous then process is not called recursively but this isn't a goal of this module. User is already warned in the header file that synchronous process may lead to recursive calls. We rather don't expect process to be synchronous because if it would be then there is no point of queuing operations.

Implementation is very complex now, with multiple places where interrupts are locked and unlocked in other functions. It's really hard to understand that. Especially that flow seems simple with only two forks: setup failure, new request pending after teardown.

Reply:
The last time I incrementally refactored this based on a belief a loop wasn't necessary, it turned out that it was necessary.  So I'm going to prioritize correctness.  We can reconsider whether it's possible to avoid looping to react to events received during unlocked periods after the solution has been shown to work correctly and be functionally complete.